src/cfg/to_structured.rs

```diff
@@ -0,0 +1,239 @@
+use std::collections::HashMap;
+
+use petgraph::{
+    algo::dominators::{simple_fast, Dominators},
+    graph::EdgeReference,
+    prelude::NodeIndex,
+    visit::EdgeRef,
+};
+
+use crate::EggCCError;
+
+use super::{
+    structured::{StructuredBlock, StructuredFunction},
+    BlockName, Branch, BranchOp, Cfg,
+};
+
+#[derive(Debug)]
+enum ContainingHistory {
+    ThenBranch,
+    LoopWithLabel(BlockName),
+    BlockFollowedBy(BlockName),
+}
+
+pub(crate) struct StructuredCfgBuilder<'a> {
+    context: Vec<ContainingHistory>,
+    postorder: HashMap<BlockName, usize>,
+    dominators: Dominators<NodeIndex>,
+    cfg: &'a Cfg,
+}
+
+impl<'a> StructuredCfgBuilder<'a> {
+    fn new(cfg: &'a Cfg) -> Self {
+        let postorder = cfg.reverse_posorder();
+        let dominators = simple_fast(&cfg.graph, cfg.entry);
+        StructuredCfgBuilder {
+            context: vec![],
+            postorder,
+            dominators,
+            cfg,
+        }
+    }
+
+    fn convert_structured(&mut self) -> Result<StructuredFunction, EggCCError> {
+        self.check_reducible()?;
```

Comment:
Not that I care too much, but is there a better way of doing this where as you build your tree you find an irreducible block and return an Err? Just thinking in the "parse don't validate" vain of things.

Reply:
I thought about this and I wasn't sure- however the check is pretty simple luckily, so I implemented it.